pkg/scalers/newrelic_scaler.go

```diff
@@ -0,0 +1,205 @@
+	if err != nil {
+		meta.region = "US"
+		newrelicLog.Info("Using default 'US' region")
+	}
+
+	if val, ok := config.TriggerMetadata[threshold]; ok && val != "" {
+		t, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %s", threshold)
+		}
+		meta.threshold = t
+	} else {
+		return nil, fmt.Errorf("missing %s value", threshold)
+	}
+
+	// If Query Return an Empty Data , shall we treat it as an error or not
+	// default is NO error is returned when query result is empty/no data
+	if val, ok := config.TriggerMetadata[noDataErr]; ok {
+		noDataErr, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, fmt.Errorf("noDataErr has invalid value")
+		}
+		meta.noDataErr = noDataErr
+	} else {
+		meta.noDataErr = false
+	}
+	meta.scalerIndex = config.ScalerIndex
+	return &meta, nil
+}
+
+func (s *newrelicScaler) IsActive(ctx context.Context) (bool, error) {
+	val, err := s.ExecuteNewRelicQuery(ctx)
+	if err != nil {
+		newrelicLog.Error(err, "error executing newrelic query")
+		return false, err
+	}
+	return val > 0, nil
+}
+
+func (s *newrelicScaler) Close(context.Context) error {
+	return nil
+}
+
+func (s *newrelicScaler) ExecuteNewRelicQuery(ctx context.Context) (float64, error) {
+	nrdbQuery := nrdb.NRQL(s.metadata.nrql)
+	resp, err := s.nrClient.Nrdb.QueryWithContext(ctx, s.metadata.account, nrdbQuery)
+	if err != nil {
+		return 0, fmt.Errorf("error running NerdGraph query %s (%s)", s.metadata.nrql, err.Error())
+	}
+	// Only use the first result from the query, as the query should not be multi row
+	for _, v := range resp.Results[0] {
+		val, ok := v.(float64)
+		if ok {
+			return val, nil
+		}
+	}
+	if s.metadata.noDataErr {
+		return 0, fmt.Errorf("query return no results %s", s.metadata.nrql)
+	}
+	return 0, nil
+}
+
+func (s *newrelicScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+	val, err := s.ExecuteNewRelicQuery(ctx)
+	if err != nil {
+		newrelicLog.Error(err, "error executing New Relic query")
+		return []external_metrics.ExternalMetricValue{}, err
+	}
+
+	metric := external_metrics.ExternalMetricValue{
+		MetricName: metricName,
+		Value:      *resource.NewQuantity(int64(val), resource.DecimalSI),
+		Timestamp:  metav1.Now(),
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
+}
+
+func (s *newrelicScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	targetMetricValue := resource.NewQuantity(int64(s.metadata.threshold), resource.DecimalSI)
+	var metricName string
+	if s.metadata.metricName == "" {
+		// TODO: Find a better way to generate a small string based on the nrql string, or maybe just use UUID?
+		hash := md5.Sum([]byte(s.metadata.nrql))
+		metricName = kedautil.NormalizeString(fmt.Sprintf("new-relic-%s", hex.EncodeToString(hash[:])))
+	} else {
+		metricName = kedautil.NormalizeString(fmt.Sprintf("new-relic-%s", s.metadata.metricName))
+	}
```

Comment:
You can drop this whole section and just use `GenerateMetricNameWithIndex(s.metadata.scalerIndex, metricName)`, it ensures that the metric name is unique.

Reply:
Ok, so we make metricName a requiered field on the metadata? is that right?

Reply:
No, it's the other way round :)

We shoulnd't ask for metricName, generating one internally is enough.